### PR TITLE
Speedup resistance_distance

### PR DESCRIPTION
**Pull request by peijenburg**
This is my first pull request ever, so there's probably a bunch of nooby errors in this PR. Any help is appreciated! Implements #6847.

Improvements to the function [resistance_distance](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.distance_measures.resistance_distance.html), primarily to speed it up, simplify its computation and to allow for the computation of the `resistance_distance` between all node pairs in the network. The main advantage of this PR is the O(N^2) speedup when computing the resistance distances of all node pairs compared to the original implementation.

Currently, the function returns the resistance distances as a dictionary (of dictionaries). Alternatively, we could also opt to return it as a numpy matrix, but I think the dictionary is preferred.

List of changes:

- Speedup and simplified computation of resistance_distance
- Allowed for computation of the resistances between all nodes. Results in a O(N^2) speedup compared to original implementation. Ensured that the script remains backwards compatible
- Handle corner case where G does not contain any nodes
- Allowed for nodeA and nodeB to be equivalent - in that case, the resistance distance is zero
- Added more test cases
- Improved references
- Added NetworkXError raises in documentation
- Mentioned that self-loops are ignored and multigraphs are contracted

### CODE REVIEW

**Comment by peijenburg** [comment]
There's this comment in the code:

> Replace with collapsing topology or approximated zero?

Does anyone know what it means?

**Comment by peijenburg** [comment]
Another question:

Currently, if the graph is disconnected, an error is returned. However, this is technically incorrect, because the effective resistance between two disconnected nodes is infinite. Unfortunately, a straightforward extension is not possible, because the algorithm fails if the graph is disconnected.

Would it be a good idea to add the following: If the graph is disconnected, find all disconnected components, and run the algorithm on each component separately. Then assemble the resistance distances by concatenating the individual components and adding np.inf for each disconnected node pair ?

Example:

1 ---- 2 ---- 3
4

rd12 = 1
rd13 = 2
rd14 = np.inf
rd23 = 1
rd24 = np.inf
rd34 = np.inf

This would require some work, but I'm wondering if it's worth the effort. What're your thoughts on this?

File: networkx/algorithms/distance_measures.py [hunk 1]
```diff
@@ -677,54 +680,64 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
...
+    if len(G) == 0:
+        msg = "Graph G must contain at least one node."
+        raise nx.NetworkXError(msg)
     if not nx.is_connected(G):
         msg = "Graph G must be strongly connected."
         raise nx.NetworkXError(msg)
-    elif nodeA not in G:
+    elif nx.is_directed(G):
+        msg = "Graph G must be undirected."
+        raise nx.NetworkXError(msg)
```
Comment (rossbar) [inline networkx/algorithms/distance_measures.py:729]: This is what the `not_implemented_for` decorator does - I'd vote to leave that as-is.

**Review by rossbar — COMMENTED** [review]
Since these changes are motivated by performance, I think it's probably worth adding benchmarks to show the performance comparison for graphs. It'd also be worth keeping an eye on the memory footprint for large graphs, since that was the reason behind using the determinant-based method rather than the pinv.

**Comment by peijenburg** [comment]
@rossbar Thanks for the feedback!

I'm not sure how to do a proper benchmark, but I have two example cases: Memory is measured using `tracemalloc` and time using the `time` module. Memory shows current vs maximum memory used and is averaged over multiple runs.
I computed the Kirchhoff index (summing the resistance distance over all node pairs) to show that both versions have the same output :)

**Graph 1: ER graph with N = 30 nodes and link-connectivity probability 0.4**

Kirchhoff index original: 74.82233343681888
Time: 2.859 s
Memory: (175112, 273088)

Kirchhoff index new: 74.82233343681865
Time: 0.008 s
Memory: (217505, 268177)

**Graph 2: ER graph with N = 100 nodes and link-connectivity probability 0.1**

Kirchhoff index original: 1208.859324125914
Time: 65.025 s
Memory: (433085, 838876)

Kirchhoff index new: 1208.8593241259148
Time: 0.083 s
Memory: (872507, 1032707)

Script used:

import networkx as nx
import numpy as np
import time
import tracemalloc

def resistance_distance2(G, nodeA=None, nodeB=None, weight=None, invert_weight=True):
    """Returns the resistance distance between every pair of nodes on graph G.

    The resistance distance between two nodes of a graph is akin to treating
    the graph as a grid of resistors with a resistance equal to the provided
    weight [1]_, [2]_.

    If weight is not provided, then a weight of 1 is used for all edges.

    If two nodes are the same, the resistance distance is zero.

    Parameters
    ----------
    G : NetworkX graph
       A graph

    nodeA : node or None, optional (default=None)
      A node within graph G.
      If None, compute resistance distance using all nodes as source nodes.

    nodeB : node or None, optional (default=None)
      A node within graph G.
      If None, compute resistance distance using all nodes as target nodes.

    weight : string or None, optional (default=None)
       The edge data key used to compute the resistance distance.
       If None, then each edge has weight 1.

    invert_weight : boolean (default=True)
        Proper calculation of resistance distance requires building the
        Laplacian matrix with the reciprocal of the weight. Not required
        if the weight is already inverted. Weight cannot be zero.

    Returns
    -------
    rd : float (if `nodeA` and `nodeB` are given)
       Resistance distance between `nodeA` and `nodeB`.
         dictionary (if `nodeA` or `nodeB` is unspecified)
       Dictionary of nodes with resistance distances as the value.

    Raises
    -------
    NetworkXNotImplemented
        If `G` is a directed graph.

    NetworkXError
        If `G` is not connected, or contains no nodes,
        or `nodeA` is not in `G` or `nodeB` is not in `G`.

    Examples
    --------
    >>> G = nx.Graph([(1, 2), (1, 3), (1, 4), (3, 4), (3, 5), (4, 5)])
    >>> round(nx.resistance_distance(G, 1, 3), 10)
    0.625

    Notes
    -----
    The implementation is based on Theorem A in [2]_. Self-loops are ignored.
    Multi-edges are contracted in one edge with weight equal to the harmonic sum of the weights.

    References
    ----------
    .. [1] Wikipedia
       "Resistance distance."
       https://en.wikipedia.org/wiki/Resistance_distance
    .. [2] D. J. Klein and M. Randic.
        Resistance distance.
        J. of Math. Chem. 12:81-95, 1993.
    """
    import numpy as np

    if len(G) == 0:
        raise nx.NetworkXError("Graph G must contain at least one node.")
    if not nx.is_connected(G):
        raise nx.NetworkXError("Graph G must be strongly connected.")
    if nodeA is not None and nodeA not in G:
        raise nx.NetworkXError("Node A is not in graph G.")
    if nodeB is not None and nodeB not in G:
        raise nx.NetworkXError("Node B is not in graph G.")

    G = G.copy()
    node_list = list(G)

    # Invert weights
    if invert_weight and weight is not None:
        if G.is_multigraph():
            for u, v, k, d in G.edges(keys=True, data=True):
                d[weight] = 1 / d[weight]
        else:
            for u, v, d in G.edges(data=True):
                d[weight] = 1 / d[weight]
    # Replace with collapsing topology or approximated zero?

    # Compute resistance distance using the Pseudo-inverse of the Laplacian
    # Self-loops are ignored
    L = nx.laplacian_matrix(G, weight=weight).todense()
    Linv = np.linalg.pinv(L, hermitian=True)

    # Return relevant distances
    if nodeA is not None and nodeB is not None:
        i = node_list.index(nodeA)
        j = node_list.index(nodeB)
        return Linv[i, i] + Linv[j, j] - Linv[i, j] - Linv[j, i]

    elif nodeA is not None:
        i = node_list.index(nodeA)
        d = {}
        for n in G:
            j = node_list.index(n)
            d[n] = Linv[i, i] + Linv[j, j] - Linv[i, j] - Linv[j, i]
        return d

    elif nodeB is not None:
        j = node_list.index(nodeB)
        d = {}
        for n in G:
            i = node_list.index(n)
            d[n] = Linv[i, i] + Linv[j, j] - Linv[i, j] - Linv[j, i]
        return d

    else:
        d = {}
        for n in G:
            i = node_list.index(n)
            d[n] = {}
            for n2 in G:
                j = node_list.index(n2)
                d[n][n2] = Linv[i, i] + Linv[j, j] - Linv[i, j] - Linv[j, i]
        return d

def GetResistancesNew(G):
    return resistance_distance2(G)

def GetResistancesOriginal(G):
    D = np.zeros((N, N))
    for i in range(N):
        for j in range(N):
            if i != j:
                D[i,j] = nx.resistance_distance(G, i, j)
            else:
                D[i,j] = 0
    return D
   

# Main
N = 100      # nodes
M = 1       # number of repeated computations
G = nx.erdos_renyi_graph(N, 0.1) # graph

t0 = time.time()
tracemalloc.start()
print('Memory', tracemalloc.get_traced_memory())
for i in range(M):
    D1 = GetResistancesOriginal(G)
    RG1 = sum(sum(D1))/2
print('Graph resistance original:', RG1)
print('Graph resistance original:', round(time.time()-t0, 3), 's')
print('Memory', tracemalloc.get_traced_memory())

t0 = time.time()
for i in range(M):
    D2 = GetResistancesNew(G)
    RG2 = sum([D2[key][key2] for key in D2 for key2 in D2[key]])/2
print('Graph resistance new:', RG2)
print('Graph resistance new:', round(time.time()-t0, 3), 's')
print('Memory', tracemalloc.get_traced_memory())
tracemalloc.stop()

**Comment by peijenburg** [comment]
The remark above is for computing **all resistances**, but what about a single resistance distance?
Answer: there is a minor (nearly negligible) speedup and a 23% memory increase, which seems reasonable to me.


**Graph 3: ER graph with N = 3000 nodes and link-connectivity probability 0.1**

Resistance original: 0.006364720759707744
Time: 7.08 s
Memory: (313,651, 374,357,969)

Resistance new: 0.0063647207597064045
Time: 6.851 s
Memory: (317,948, 461,696,465)

**Comment by dschult** [comment]
Thank you for the timing and memory comparisons. And good to also include the comparison for a single pair of nodes.

TLDR: I prefer this pseudo-inverse method over the determinant method.  
The time gain and flexibility of computing for multiple pairs is worth the rather small impact on memory.

Details:
Comparison could be tricky:
- The current version uses determinants to find the result for a single pair of nodes. (less memory, more time)
- This PR solves the whole circuit/graph essentially finding the result for all pairs of nodes. (less time, more memory)

The first method saves some memory relative to the second, but at a large computational cost. There is a general rule of computational liner algebra that determinants are often great for proofs but often bad for computations. In this case, it is a further disadvantage that the determinant route doesn't even give results for all pairs -- only for one pair. You must compute the determinants for "L without node a" and "L without nodes a and b". So to do all nodes you must compute lots of determinants. I looked at the determinant method to figure out whether it might be converted into an all-pairs method, but I don't see a good way to do that.

Looking at the time/memory comparison for a single pair of nodes (the environment favoring the current code), I don't believe the memory cost of the PR methodology could be more than twice that of the original (scales nicely). The time advantage of the PR is present, but not large in this setting. But the memory cost is also not large. Indeed the memory savings might be saved elsewhere (e.g. by eliminating the copy of the original graph).  But the advantage of the PR is that you've basically solved the entire problem to get the first pair result. This leads to very large time savings very quickly.  

Perhaps we could make up for that memory impact by e.g. rewriting `nx.laplacian_matrix` to allow a function `weight` instead of forcing a string attribute (see #4085). That would eliminate the need for copying the graph. But I haven't checked that copying the graph is significant in this method.

**Review by dschult — APPROVED** [review]
Thanks! LGTM

File: networkx/algorithms/tests/test_distance_measures.py [hunk 2]
```diff
@@ -369,22 +390,13 @@ def test_resistance_distance_div0(self):
...
+        assert nx.resistance_distance(self.G, 1, 1) == 0
 
-    def test_resistance_distance_nodeA_not_in_graph(self):
-        with pytest.raises(nx.NetworkXError):
-            nx.resistance_distance(self.G, 9, 1)
-
-    def test_resistance_distance_nodeB_not_in_graph(self):
-        with pytest.raises(nx.NetworkXError):
-            nx.resistance_distance(self.G, 1, 9)
+    def test_resistance_distance_all(self):
```
Comment (rossbar) [inline networkx/algorithms/tests/test_distance_measures.py:396]: Since the semantics of `nodeA` and `nodeB` have been altered, we'll also need additional tests for the new behavior; i.e. a case when `nodeA` is not `None` and another when `nodeB` is no `None`.

**Review by rossbar — APPROVED** [review]
LGTM, thanks @peijenburg !
